### PR TITLE
identity: prepare Credentials API for Spire support

### PR DESCRIPTION
**Pull request by zaharidichev**
This PR prepares the Credentials API for integrating with SPIRE more easily.
Notable changes include: 

- the `gen_certificate_signing_request` is not part of the concrete identity service client rather than the `Credentials` trait
- CSR, service account tokens, etc are supposed to be a concern of the concrete LinkedIn identity client implementation, so the Store and Receiver do not know about these
- the fn `set_certificate` method now takes the cert key as well, because in alternative certificate provider implementations (e.g. Spire), we cannot assume that the certificate key is known upfront or is static.

### CODE REVIEW

File: linkerd/meshtls/rustls/src/creds/store.rs [hunk 1]
```diff
@@ -156,9 +145,11 @@ impl id::Credentials for Store {
         // Use the client's verifier to validate the certificate for our local name.
         self.validate(&chain)?;
 
+        let key = EcdsaKeyPair::from_pkcs8(SIGNATURE_ALG_RING_SIGNING, &key).map_err(InvalidKey)?;
+
         let resolver = Arc::new(CertResolver(Arc::new(rustls::sign::CertifiedKey::new(
             chain,
-            Arc::new(Key(self.key.clone())),
+            Arc::new(Key(Arc::new(key))),
```
[the remark this answers is not in the document]
Comment (zaharidichev) [reply linkerd/meshtls/rustls/src/creds/store.rs:152]: Seems like we can remove the inner Arc, but the outer is needed by the API.

File: linkerd/app/src/env.rs [hunk 2]
```diff
@@ -1249,12 +1249,13 @@ pub fn parse_identity_config<S: Strings>(
                 token,
                 min_refresh: min_refresh.unwrap_or(DEFAULT_IDENTITY_MIN_REFRESH),
                 max_refresh: max_refresh.unwrap_or(DEFAULT_IDENTITY_MAX_REFRESH),
+                key_pkcs8: key?,
```
[the remark this answers is not in the document]
Comment (zaharidichev) [reply linkerd/app/src/env.rs:1252]: I have just added a custom impl of `Debug` for the new Documents type. Would that wor?  So far, this has been the approach. Is that sufficient?

[the remark this answers is not in the document]
Comment (zaharidichev) [reply linkerd/app/src/env.rs:1252]: @olix0r  Just to elaborate a bit on the motivation behind this shuffling. The end state that I envision for this identity::Config type kind lof looks like this: 

Goring from: 
```rust
#[derive(Clone, Debug)]
pub struct Config {
    pub control: control::Config,
    pub certify: certify::Config,
    pub params: TlsParams,
}
```

To: 

```rust
pub enum Provider {
    ControlPlane {
        control: control::Config,
        certify: certify::Config,
    },
    Spire {
        socket_path: Arc<String>,
    },
}

pub struct Config {
    pub provider: Provider,
    pub params: TlsParams,
}
```

And then in Config::build, we construct the identity daemon Task to be either a spire or a control plane daemon, based on the variant of the enum.

**Comment by olix0r** [comment]
@zaharidichev here are some small suggestions that I have: https://github.com/linkerd/linkerd2-proxy/compare/zd/prepare-credentials-api-for-spire...ver/paranoia?expand=1

> This change tightens up scope/visibility of private key material so that it only ever exists in certify.rs (before being passed to the identity store).
>
> * The fields on `Documents` are made private so they can't be read from other modules.
> * To accomodate this, the loading logic is moved from env.rs to certify.rs
> * Documents is now stored in an Arc so it can be cloned cheaply (Configs should be cheap to clone, generally).
> * Documents Debug impl is updated to use the std library debug helpers with "..." in place of field data.

Note that I had to push your branch to this fork to get this diff to render properly, so you'll probably want to merge ver/paranoia back outside of the UI.

**Comment by zaharidichev** [comment]
@olix0r thanks a bunch for the feedback!

File: linkerd/app/src/identity.rs [hunk 3]
```diff
@@ -100,28 +97,19 @@ impl Config {
...
         let _ = self.tx.send(true);
         Ok(())
     }
 }
 
-// === impl Documents ===
+// === impl TlsParams ===
 
-impl std::fmt::Debug for Documents {
+impl std::fmt::Debug for TlsParams {
```
Comment (olix0r) [inline linkerd/app/src/identity.rs:109]: FWIW it looks like we could safely derive(Debug) for this now

File: linkerd/meshtls/rustls/src/creds/store.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (olix0r) [reply linkerd/meshtls/rustls/src/creds/store.rs:152]: we probably don't need to derive Clone for our type since we're always passing it into an Arc<Key>

**Review by olix0r — APPROVED** [review]

**Pull request merged**